salt/pwyf-dqt.sls

```diff
@@ -0,0 +1,168 @@
+
+##### System install
+
+include:
+  - core
+  - apache
+  - uwsgi
+#   - letsencrypt
+
+# uWSGI from data-quality-tester > requirements.txt has a C
+# extension not sure if this is needed given we install the pkg
+# uwsgi-plugin-python3 but it is in the current requirements so we
+# need to make sure it is installable
+
+pwyf-dqt-deps:
+  pkg.installed:
+    - pkgs:
+      - virtualenv
+      - libapache2-mod-proxy-uwsgi
+      - uwsgi-plugin-python3
+      - redis-server
+      # See uWSGI comment ^
+      - build-essential
+      - python3-dev
+    - watch_in:
+      - service: apache2
+      - service: uwsgi
+
+  apache_module.enabled:
+    - name: proxy proxy_uwsgi
+    - watch_in:
+      - service: apache2
+      - service: uwsgi
+
+# redis should start up after being installed but this will verify
+
+redis-server:
+  pkg:
+    - installed
+  service:
+    - running
+    - enable: True
+    - reload: True
+
+{{ createuser(pillar.pwyf_dqt.user) }}
+
+##### Git checkout 
+
+/home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}:
+  git.latest:
+    - name: {{ pillar.pwyf_dqt.git_url }}
+    - rev: {{ pillar.pwyf_dqt.branch }}
+    - user: {{ pillar.pwyf_dqt.user }}
+    - target: /home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}
+    - force_fetch: True
+    - force_reset: True
+    - submodules: True
+    - require:
+      - pkg: git
+
+##### Install python deps in virtualenv
+
+/home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/.ve:
+  virtualenv.managed:
+    - python: /usr/bin/python3
+    - user: {{ pillar.pwyf_dqt.user }}
+    - system_site_packages: False
+    - requirements: /home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/requirements.txt
+    - require:
+      - git: /home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}
+
+##### Flask app setup
+
+setup_dqt:
+  cmd.run:
+    - runas: {{ pillar.pwyf_dqt.user }}
+    - env:
+      - FLASK_APP: 'DataQualityTester/__init__.py'
+      - FLASK_ENV: 'development'
+    - cwd: /home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/
+    - name: source ./.ve/bin/activate; cp config.py.tmpl config.py; flask db upgrade; flask assets build
+
+## Flask config
+/home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/config.py:
+ file.append:
+    - text:
+      - "    SQLALCHEMY_DATABASE_URI='sqlite:////home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/db.sqlite3'"
+      - "    SECRET_KEY='{{ pillar.pwyf_dqt_private.secret_key }}'"
+    - require:
+      - setup_dqt
+
+
+
+{{ pillar.pwyf_dqt.static_dir }}:
+  file.directory:
+    - user: {{ pillar.pwyf_dqt.user }}
+    - group: www-data
+    - mode: 755
+    - makedirs: True
+
+setup_dqt_assets:
+  cmd.run:
+    - name: cp -r /home/{{ pillar.pwyf_dqt.user }}/{{ pillar.pwyf_dqt.checkout_dir }}/DataQualityTester/static/* {{ pillar.pwyf_dqt.static_dir }} ; chown -R www-data.www-data {{ pillar.pwyf_dqt.static_dir }}
```

Comment:
One consequence of doing it this way is that static dir is never cleared out; if a file that is put in git is later  removed it will still be available on the web server. That probably doesn't matter tho .... ? (Maybe a comment to say we have decided it's not important?)

Reply:
Hmm yeah I'll have think, could just clear it before and then rysnc -r --delete or something. I should probably also add a watcher so that it is run if the git repo is updated.